Add fiber summary tooltip to devtools profiling

M-Izadmehr opened the pull request:
## Summary
The purpose of this MR is to address #18014 (when working on a large project/component tree it can be difficult to understand why a large subtree is re-rendering, as a result adding a tooltip fo devtools profiler, which makes the process of debugging and inspecting elements easier.)

As a result, in really small components we don't need to click on them to see the details in sidebar, now it is possible to see the summary in the tooltip.




## Results

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/28848972/74614072-08155400-5115-11ea-8d19-7ab3d27b9b0a.gif)
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/28848972/74614074-09468100-5115-11ea-8c87-c224d229ef15.gif)

bvaughn commented:
Screencast of this looks great!

I'll review it soon.

strdr4605 commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js:
```diff
@@ -52,6 +77,8 @@ function CommitRankedListItem({data, index, style}: Props) {
       label={node.label}
       onClick={handleClick}
+      onMouseOver={event => handleMouseOver(event, node)}
+      onMouseOut={event => handleMouseOut(event)}
```
Maybe
```suggestion
      onMouseOut={handleMouseOut}
```

strdr4605 commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js:
```diff
@@ -52,6 +77,8 @@ function CommitRankedListItem({data, index, style}: Props) {
       key={node.id}
       label={node.label}
       onClick={handleClick}
+      onMouseOver={event => handleMouseOver(event, node)}
```
As `node` can be accessed from closure, maybe 
```suggestion
      onMouseOver={handleMouseOver}
```

strdr4605 commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js:
```diff
@@ -38,6 +46,23 @@ function CommitRankedListItem({data, index, style}: Props) {
+      hoverFiber({id, name});
+    },
+    [hoverFiber],
+  );
```
As`nodeData` can be accessed from closure, maybe
```js
  const handleMouseOver = useCallback(
    (event: SyntheticMouseEvent<*>) => {
      const {id, name} = node;
      event.stopPropagation();
      hoverFiber({id, name});
    },
    [hoverFiber, node],
  );
```

M-Izadmehr replied:
Makes sense, I removed it as param and used the value from closure with a simple callback

strdr4605 commented on packages/react-devtools-shared/src/devtools/views/Profiler/HoveredFiberInfo.js:
```diff
@@ -0,0 +1,78 @@
+      );
+      break;
+    }
+  }
```
Looks like this loop only finds one element, so why the `listItems` array is needed?
Maybe something like this?
```js
  const item = useMemo(() => {
    const commitIndices = profilingCache.getFiberCommits({
      fiberID: ((id: any): number),
      rootID: ((rootID: any): number),
    });
    const isSelectedCommitIndexInCommitIndices = commitIndices.some((commitIndex) => selectedCommitIndex === commitIndex);
    if (isSelectedCommitIndexInCommitIndices) {
      const {duration, timestamp} = profilerStore.getCommitData(
        ((rootID: any): number),
        selectedCommitIndex,
      );
      return (
        <div key={commitIndex} className={styles.CurrentCommit}>
          {formatTime(timestamp)}s for {formatDuration(duration)}ms
        </div>
      );
    } else {
      return null;
    }
  }, [selectedCommitIndex]);
```
Also `item` is not a good name so should think on renaming it.

M-Izadmehr replied:
Fixed

bvaughn commented:
Need to run into a meeting now so a full PR is pending, but I noticed the tooltip position flickers when I did a quick test, e.g.:
![flickerKapture 2020-02-18 at 10 25 27](https://user-images.githubusercontent.com/29597/74765764-03e25700-5239-11ea-9349-781f5aa4a13f.gif)

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Components/Tooltip.js:
```diff
@@ -0,0 +1,51 @@
+  const [tooltipState, setTooltipState] = useState(initialTooltipState);
+  const tooltipRef = useSmartTooltip(tooltipState);
+
+  const onMouseMove = useCallback(
```
`useCallback` doesn't do anything in this case, so we can remove it :)

M-Izadmehr replied:
Done

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/ChartNode.js:
```diff
@@ -41,12 +45,13 @@ export default function ChartNode({
         fill={color}
         onClick={onClick}
+        onMouseOver={onMouseOver}
+        onMouseOut={onMouseOut}
```
nit: Not sure it matters in this case (SVG rect) but in general, I'd prefer using mouse-enter and mouse-leave rather than mouse-over and mouse-out

M-Izadmehr replied:
Yeah, agree, I changed them to `onMouseLeave/Enter`, I think we also don't need to `stopPropagation` anymore, I removed them too:)

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitFlamegraphListItem.js:
```diff
@@ -104,6 +124,8 @@ function CommitFlamegraphListItem({data, index, style}: Props) {
             label={label}
             onClick={event => handleClick(event, id, name)}
+            onMouseOver={event => handleMouseOver(event, chartNode)}
+            onMouseOut={event => handleMouseOut(event)}
```
Same nit as above (enter & leave)

M-Izadmehr replied:
Done

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitFlamegraphListItem.js:
```diff
@@ -43,6 +46,23 @@ function CommitFlamegraphListItem({data, index, style}: Props) {
     [selectFiber],
   );
 
+  const handleMouseOver = useCallback(
```
These `useCallback` for mouse handlers can be removed :)

M-Izadmehr replied:
Done!

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js:
```diff
@@ -38,6 +46,23 @@ function CommitRankedListItem({data, index, style}: Props) {
     [node, selectFiber],
   );
 
+  const handleMouseOver = useCallback(
```
`useCallback` can be removed here too.

M-Izadmehr replied:
Done.

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js:
```diff
@@ -38,6 +46,23 @@ function CommitRankedListItem({data, index, style}: Props) {
+    [hoverFiber],
+  );
+
+  const handleMouseOut = useCallback(
```
And here too :)

M-Izadmehr replied:
Done.

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/HoveredFiberInfo.js:
```diff
@@ -0,0 +1,78 @@
+        commitIndex,
+      );
+
+      listItems.push(
```
This doesn't need to be an array since it will only ever find 0 or 1 matches.

M-Izadmehr replied:
I had doubts about whether to change it to filter, or keep it as for, as I checked I saw that we are banning `for...in` in eslints, also I never saw a case where we use `filter/map`, is there a particular reason for that? :)

M-Izadmehr replied:
Done by the way!

bvaughn commented on packages/react-devtools-shared/src/devtools/views/Profiler/HoveredFiberInfo.js:
```diff
@@ -0,0 +1,78 @@
+        )}
+        {listItems.length === 0 && (
+          <div>Did not render during this profiling session.</div>
+        )}
```
If you replaced the list above with a single item, this could be re-written more simply as e.g. `{matchingListItem || <div>Did not render during this profiling session.</div>}`

M-Izadmehr commented:
@bvaughn @strdr4605 Thanks for great comments, I will try to check them and also the flicker issue and get back to you:)

bvaughn commented:
Thank YOU for the great work so far :) Much appreciated

M-Izadmehr commented:
Yeah, good catch, the flicker was due to my last commit, initially, I was doing `onMouseMove` on every movement of the mouse and it caused two updates (1. due to setting state of tooltip position, 2. sync useLayoutEffect), which would result in re-render of the whole graph on change. The cause of this issue was that in my last commit I added a condition to *"only listen for movements only if there is an element hovered"* (which was not the best decision:) )
Now, I removed the `useEffectLayout`, and also `tooltipState`, and as a result, the `children` (the whole graph) never re-renders because of mouse movement, and flicker is also fixed:)
![ezgif com-video-to-gif (2)](https://user-images.githubusercontent.com/28848972/74794839-a4318d80-52c4-11ea-9b26-a1a43909f3f8.gif)

bvaughn commented:
Ooh, thanks for the quick turnaround! Let me take another look!

strdr4605 commented on packages/react-devtools-shared/src/devtools/views/Profiler/CommitFlamegraphListItem.js:
```diff
@@ -43,6 +46,15 @@ function CommitFlamegraphListItem({data, index, style}: Props) {
+
+  const handleMouseLeave = () => {
+    hoverFiber(null);
+  };
```
Looks like this is same functions as in https://github.com/facebook/react/blob/0bd6e07a970fc711855b184ca222ee6d1897ade3/packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js#L48

So for consistency use variant in https://github.com/facebook/react/blob/0bd6e07a970fc711855b184ca222ee6d1897ade3/packages/react-devtools-shared/src/devtools/views/Profiler/CommitRankedListItem.js#L48

M-Izadmehr replied:
Thanks, I removed this useCallback too:)

bvaughn commented:
FYI this change has just been released to Chrome/Firefox/Edge/NPM in] version 4.5.

https://github.com/facebook/react/blob/master/packages/react-devtools/CHANGELOG.md#450-march-3-2020

Thank you!
